### ISSUE TITLE
Heterogeneous GASNET_DOMAIN_COUNT value causes hangs

### ISSUE THREAD

**Issue by ronawho** [chapel-lang/chapel#11621]
Under gasnet-aries we set `GASNET_DOMAIN_COUNT` to `<numCores+1>`. We've been seeing intermittent hangs on a heterogenous system, which we've narrowed down to a single job getting different node types. As a specific example a job on a 28-core BW and 68-core KNL will hang. However, not all combinations seem to hang as a job on 10-core IV and 20-core IV seems to work.

Here's an awful patch that I've been using to reproduce. It sets `GASNET_DOMAIN_COUNT` to `SLURM_NODEID+1`

```diff
diff --git a/runtime/src/comm/gasnet/comm-gasnet.c b/runtime/src/comm/gasnet/comm-gasnet.c
index 2b1e0c6f25..76971f6014 100644
--- a/runtime/src/comm/gasnet/comm-gasnet.c
+++ b/runtime/src/comm/gasnet/comm-gasnet.c
@@ -766,12 +766,13 @@ static void set_max_segsize() {
   }
 }

+#include "chpl-env.h"
 static void set_num_comm_domains() {
 #if defined(GASNET_CONDUIT_GEMINI) || defined(GASNET_CONDUIT_ARIES)
   char num_cpus_val[22]; // big enough for an unsigned 64-bit quantity
   int num_cpus;

-  num_cpus = chpl_topo_getNumCPUsPhysical(true) + 1;
+  num_cpus = chpl_env_str_to_int("SLURM_NODEID", getenv("SLURM_NODEID"), 0) + 1;

   snprintf(num_cpus_val, sizeof(num_cpus_val), "%d", num_cpus);
   if (setenv("GASNET_DOMAIN_COUNT", num_cpus_val, 0) != 0) {
```

This fails for me with 2 or more locales.

**Comment by bonachea** [comment]
@ronawho I believe this is fundamental to the way the multi-domain support works, requiring the same number of underlying domain creations on every node (each domain is an isolated GNI-level virtual network across all the processes). This envvar should probably be documented as requiring the same value on all nodes, although I see that documentation is currently lacking.

I suggest setting the value to MAX(num_cores) across all nodes in the job.

**Comment by ronawho** [comment]
Hmm, interesting. I don't think we have a programatic way to determine the max number of cores before initializing gasnet.

For our `comm=ugni` we do an out-of-band PMI_Allgather to share and compute the max number of comm domains, but I don't think we have a way of doing that under gasnet.

**Comment by ronawho** [comment]
I'm not very familiar with GASNet core code, but the following seems to work for me for broadcasting and finding the max number of comm domains:

```diff
diff --git a/third-party/gasnet/gasnet-src/gemini-conduit/gasnet_gemini.c b/third-party/gasnet/gasnet-src/gemini-conduit/gasnet_gemini.c
index 3356fde58e..16dbdf8d1e 100644
--- a/third-party/gasnet/gasnet-src/gemini-conduit/gasnet_gemini.c
+++ b/third-party/gasnet/gasnet-src/gemini-conduit/gasnet_gemini.c
@@ -894,6 +894,16 @@ uintptr_t gasnetc_init_messaging(void)
                GASNETC_DOMAIN_COUNT_DEFAULT,0);
   gasnetc_poll_am_domain_mask = gasneti_getenv_int_withdefault("GASNET_AM_DOMAIN_POLL_MASK",
                GASNETC_AM_DOMAIN_POLL_MASK_DEFAULT,0);
+
+  unsigned int *all_gasnetc_domain_count = gasneti_malloc(gasneti_nodes * sizeof(unsigned int));
+  gasneti_spawner->Exchange(&gasnetc_domain_count, sizeof(gasnetc_domain_count), all_gasnetc_domain_count);
+  for (i = 0; i < gasneti_nodes; i += 1) {
+    if (all_gasnetc_domain_count[i] > gasnetc_domain_count) {
+      gasnetc_domain_count = all_gasnetc_domain_count[i];
+    }
+  }
+  gasneti_free(all_gasnetc_domain_count);
+
   if (gasnetc_poll_am_domain_mask) {
     for (i=2; i<gasnetc_domain_count; i<<=1) {
       gasnetc_poll_am_domain_mask = (gasnetc_poll_am_domain_mask << 1) | 1;
```

**Comment by bonachea** [comment]
Thanks @ronawho - I believe your suggested patch is _probably_ correct, although I have not tested it myself. In any case I agree aries-conduit multi-domain support needs better support for heterogeneous nodes, and this is probably something GASNet should handle automatically, if possible.

Paul is at SC18 right now, but we'll discuss this and get back to you next week (if not sooner). My main concern is whether there are any negative implications of creating domains that some processes would never poll (in which case the reduction may need to be a MIN operation instead of a MAX).

**Comment by PHHargrove** [comment]
I do not know off the top of my head the entire answer to Dan's concern above about negative implications of creating more domains than are used.  The one consequence that comes to mind is that if there are more than 123(ish) domains created then we need to use FMA sharing, with a corresponding performance penalty.

Threads are bound to domains round-robin.  So, the use of a MIN is safe if that is the outcome.

I am not even 100% sure the "diagnosis" is correct, given the following snippet of code (the comments actually):
```
#if 0 /* The following is the "proper" binding to "same-domain" peers... */
   status = GNI_EpBind(DOMAIN_SPECIFIC_VAL(peer_data[i].ep_handle),
                       gasnetc_cdom_data[0].peer_data[i].nic_addr,
                       i + GASNETC_DIDX * gasneti_nodes);
#else /* ...but bind to remote domain=0 is insensitive to having fewer domains on some nodes */
   status = GNI_EpBind(DOMAIN_SPECIFIC_VAL(peer_data[i].ep_handle),
                       gasnetc_cdom_data[0].peer_data[i].nic_addr,
                       i);
#endif
```
It looks like there we are trying *not* to create an "isolated virtual network", but the same may not be true in other parts of the code.  So, I am not prepared to give my final answer until I've had a chance to reproduce the problem and track down a root cause.

@ronawho or @bonachea: could one of you please transcribe the key points to a GASNet-EX Bugzilla entry for me?

I do believe the patch to take a MAX is safe.  So, if it is also effective then it should be fine as a short-term solution until I've identified a root cause.  Ideally I should be able to support heterogenous domain count, but I cannot promise that until I've refreshed my understanding of this code.

BTW: GASNet-EX support for multiple endpoints will definitely not have this equality restriction.

**Comment by PHHargrove** [comment]
Regarding the cost of the unused domains due to the MAX:

As mentioned before, the allocation of more GNI-level Cdm's will potentially push the implementation into FMA sharing mode.

We also create additional GNI-level Cq's and Ep's.
Other than the memory they consume, I am not aware of any associated cost (nothing analogous to the FMA-sharing overhead).

In addition, I find 416KB of pinned memory allocated for each domain.  
This will consume physical page frames for each "extra" domain, but will not get used.  
However, with this memory coming from 2MB huge pages, it may or may not be visible.

I am hoping I will have a solution which will avoid allocation of the GNI-level resources.

**Comment by PHHargrove** [comment]
> I am hoping I will have a solution which will avoid allocation of the GNI-level resources.

A proposed fix appears in [GASNet PR #243](https://bitbucket.org/berkeleylab/gasnet/pull-requests/243).  

@ronawho, please verify that commit works for you (via the PR interface if possible).  
If/when merged, I will provide a final commit hash for you include in Chapel.

**Comment by ronawho** [comment]
Confirmed that it works for me. Works for my simple reproducer and running on all nodes in a system that has a bunch of different node types (BW28, IV20, IV24, HW24, HW32, KNL64, and KNL68)

**Comment by PHHargrove** [comment]
Resolved in GASNet-1 at commit [8b551f44](https://bitbucket.org/PHHargrove/gasnet-public/commits/8b551f44)  
Resolved in GASNet-EX at commit [5f5199a2](https://bitbucket.org/berkeleylab/gasnet-ex/commits/5f5199a2) (though that repo is not yet public).

**Comment by ronawho** [comment]
Cherry picked in https://github.com/chapel-lang/chapel/pull/11662

As always,  thanks for the fast turnaround!
